### PR TITLE
Key extraction for CryptGenKey

### PR DESCRIPTION
**Pull request by sasza8**
Custom hook for CryptGenKey from cryptoapi allowing to extract generated key bytes

### CODE REVIEW

**Comment by tklengyel** [comment]
So I'm wondering whether this should be a separate plugin from memdump. The functionality this provides certainly seems distinct enough to do that. Question is how much it depends on the plumbing of memdump and whether that could be abstracted out to either libdrakvuf or a common plugin interface plugins could opt into.

**Comment by icedevml** [comment]
@tklengyel We are aware that this problem araises with every non-memdumping-related PR to `memdump`. I think it is around the time to split `memdump` into `memdump` and `usermode`(?), but then we need to move a lot of stuff into libdrakvuf and perhaps this is going to be a large change.

Maybe let's merge this, but we will open a ticket for splitting `memdump` and I will take care of it within few days?

**Comment by tklengyel** [comment]
> but then we need to move a lot of stuff into libdrakvuf and perhaps this is going to be a large change

As long as the code isn't using too much C++ specific stuff that move should be quite straight forward. And don't worry about bloating libdrakvuf, it's purpose is to house common functionality like this.

> Maybe let's merge this, but we will open a ticket for splitting memdump and I will take care of it within few days?

SGTM

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by icedevml** [comment]
Related #787

File: src/plugins/memdump/crypto.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+#include <inttypes.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <libdrakvuf/private.h>
+#include <assert.h>
+#include <vector>
+#include <map>
+#include <string>
+
+#define MAGIC_PTR_XOR_VALUE 0xE35A172C
```
Comment (tklengyel) [inline src/plugins/memdump/crypto.h]: Perhaps a comment to explain where the magic value comes from?

File: src/plugins/memdump/misc.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef DRAKMON_MISC_H
+#define DRAKMON_MISC_H
+
+
+bool is_wow64(drakvuf_t drakvuf, drakvuf_trap_info_t* info);
+addr_t get_function_argument(drakvuf_t drakvuf, drakvuf_trap_info_t* info, int narg);
```
Comment (tklengyel) [inline src/plugins/memdump/misc.h]: Both of these functions could be very easily moved to libdrakvuf right away, albeit get_function_argument would have to do things differently depending on the OS. Linux has 6 arguments in the different registers vs 4 in Windows.

File: src/plugins/memdump/crypto.h (hunk 1 above)
Comment (icedevml) [reply src/plugins/memdump/crypto.h]: done

Comment (sasza8) [reply src/plugins/memdump/crypto.h]: nobody knows ¯\_(ツ)_/¯
I just reversed the DLL
maybe we should ask MS developers who wrote this code 30 years ago, if they're still alive

Comment (tklengyel) [reply src/plugins/memdump/crypto.h]: FYI: https://github.com/iSECPartners/jailbreak/blob/master/jailbreakhook/jailbreakhook.h#L28 Do you need perhaps both magic values as seen there?

Comment (sasza8) [reply src/plugins/memdump/crypto.h]: For now it only works for 32-bits WOW binaries, implementing support for 64 needs a little more effort and will be developed in future - and yes, then we'll need that value

Comment (tklengyel) [reply src/plugins/memdump/crypto.h]: OK, so add a TODO comment to highlight that and then we can merge it

Comment (sasza8) [reply src/plugins/memdump/crypto.h]: there already was implicit TODO in hook function, but added comment either way

File: src/plugins/memdump/misc.h (hunk 2 above)
Comment (sasza8) [reply src/plugins/memdump/misc.h]: moved

**Comment by sasza8** [comment]
@tklengyel CI failed, could you paste logs here?

**Comment by tklengyel** [comment]
It's unrelated to this PR.

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Pull request merged**